File: nng_comm/tests/test_cross_lang.cpp
#include "nng_comm/nng_comm.hpp"
#include <iostream>
#include <thread>
#include <chrono>
#include <cassert>
#include <cstring>

using namespace nng_comm;

// C++ 发布，等待对端订阅后发送消息
void run_publisher(const std::string& address) {
    Publisher pub;
    assert(pub.init(address));
    std::cout << "[C++ Publisher] listening on " << address << std::endl;

    // 等待订阅端连接
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    for (int i = 0; i < 20; i++) {
        nlohmann::json msg = {
            {"source", "cpp"},
            {"type", "cross_lang_test"},
            {"seq", i},
            {"value", 42}
        };
        Message::add_timestamp(msg);
        pub.publish(msg);
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // 等待对端处理完毕
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    pub.close();
    std::cout << "[C++ Publisher] done" << std::endl;
}

// C++ 订阅，接收消息并验证
void run_subscriber(const std::string& address) {
    Subscriber sub;
    assert(sub.init(address));
    std::cout << "[C++ Subscriber] dialing " << address << std::endl;

    nlohmann::json msg;
    bool received = false;
    for (int attempt = 0; attempt < 30 && !received; attempt++) {
        received = sub.receive(msg, 500);
    }

    assert(received);
    assert(msg["source"] == "python");
    assert(msg["type"] == "cross_lang_test");
    std::cout << "[C++ Subscriber] received: " << msg.dump() << std::endl;

    sub.close();
    std::cout << "[C++ Subscriber] PASSED" << std::endl;
}

// C++ 提供服务（Replier）
void run_replier(const std::string& address) {
    Replier rep;
    assert(rep.init(address));

    rep.set_handler([](const nlohmann::json& req) -> nlohmann::json {
        std::cout << "[C++ Replier] received: " << req.dump() << std::endl;
        if (req["command"] == "add") {
            int result = req["a"].get<int>() + req["b"].get<int>();
            return {{"status", "ok"}, {"result", result}, {"server", "cpp"}};
        }
        return {{"status", "error"}, {"message", "unknown command"}};
    });

    rep.start();
    std::cout << "[C++ Replier] listening on " << address << std::endl;

    // 运行一段时间等待请求
    std::this_thread::sleep_for(std::chrono::seconds(5));

    rep.stop();
    rep.close();
    std::cout << "[C++ Replier] done" << std::endl;
}

// C++ 请求服务（Requester）
void run_requester(const std::string& address) {
    // 等待服务端启动
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    Requester req;
    assert(req.init(address));
    std::cout << "[C++ Requester] dialing " << address << std::endl;

    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    nlohmann::json request = {
        {"command", "add"},
        {"a", 100},
        {"b", 200}
    };

    nlohmann::json reply;
    assert(req.request(request, reply, 5000));
    std::cout << "[C++ Requester] reply: " << reply.dump() << std::endl;

    assert(reply["status"] == "ok");
    assert(reply["result"] == 300);
    assert(reply["server"] == "python");

    req.close();
    std::cout << "[C++ Requester] PASSED" << std::endl;
}

void print_usage(const char* prog) {
    std::cerr << "Usage: " << prog << " <mode> <address>" << std::endl;
    std::cerr << "  mode: pub | sub | rep | req" << std::endl;
    std::cerr << "  address: e.g. tcp://127.0.0.1:15580" << std::endl;
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        print_usage(argv[0]);
        return 1;
    }

    std::string mode = argv[1];
    std::string address = argv[2];

    try {
        if (mode == "pub") {
            run_publisher(address);
        } else if (mode == "sub") {
            run_subscriber(address);
        } else if (mode == "rep") {
            run_replier(address);
        } else if (mode == "req") {
            run_requester(address);
        } else {
            print_usage(argv[0]);
            return 1;
        }
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "FAILED: " << e.what() << std::endl;
        return 1;
    }
}

File: nng_comm/tests/test_cross_lang.py
#!/usr/bin/env python3
"""Cross-language test: Python side.

Usage:
    python test_cross_lang.py <mode> <address>
    mode: pub | sub | rep | req
    address: e.g. tcp://127.0.0.1:15580
"""
import sys
import os
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from scripts import Publisher, Subscriber, Requester, Replier, Message


def run_publisher(address: str):
    pub = Publisher()
    assert pub.init(address), "Publisher init failed"
    print(f"[Python Publisher] listening on {address}")

    # 等待订阅端连接
    time.sleep(0.5)

    for i in range(20):
        msg = {"source": "python", "type": "cross_lang_test", "seq": i, "value": 42}
        Message.add_timestamp(msg)
        pub.publish(msg)
        time.sleep(0.1)

    # 等待对端处理完毕
    time.sleep(0.5)
    pub.close()
    print("[Python Publisher] done")


def run_subscriber(address: str):
    sub = Subscriber()
    assert sub.init(address), "Subscriber init failed"
    print(f"[Python Subscriber] dialing {address}")

    received = None
    for _ in range(30):
        received = sub.receive(timeout_ms=500)
        if received is not None:
            break

    assert received is not None, "No message received"
    assert received["source"] == "cpp", f"Expected source 'cpp', got '{received['source']}'"
    assert received["type"] == "cross_lang_test"
    print(f"[Python Subscriber] received: {received}")

    sub.close()
    print("[Python Subscriber] PASSED")


def run_replier(address: str):
    def handler(req):
        print(f"[Python Replier] received: {req}")
        if req.get("command") == "add":
            result = req.get("a", 0) + req.get("b", 0)
            return {"status": "ok", "result": result, "server": "python"}
        return {"status": "error", "message": "unknown command"}

    rep = Replier()
    assert rep.init(address), "Replier init failed"
    rep.set_handler(handler)
    rep.start()
    print(f"[Python Replier] listening on {address}")

    # 运行一段时间等待请求
    time.sleep(5)

    rep.stop()
    rep.close()
    print("[Python Replier] done")


def run_requester(address: str):
    # 等待服务端启动
    time.sleep(0.5)

    req = Requester()
    assert req.init(address), "Requester init failed"
    print(f"[Python Requester] dialing {address}")

    time.sleep(0.2)

    request = {"command": "add", "a": 100, "b": 200}
    reply = req.request(request, timeout_ms=5000)
    print(f"[Python Requester] reply: {reply}")

    assert reply is not None, "No reply received"
    assert reply["status"] == "ok"
    assert reply["result"] == 300
    assert reply["server"] == "cpp"

    req.close()
    print("[Python Requester] PASSED")


def main():
    if len(sys.argv) != 3:
        print(__doc__)
        sys.exit(1)

    mode = sys.argv[1]
    address = sys.argv[2]

    runners = {
        "pub": run_publisher,
        "sub": run_subscriber,
        "rep": run_replier,
        "req": run_requester,
    }

    if mode not in runners:
        print(f"Unknown mode: {mode}")
        print(__doc__)
        sys.exit(1)

    runners[mode](address)


if __name__ == "__main__":
    main()
